File: java/convert_sorted_array_to_binary_search_tree.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.Stack;
import java.util.PriorityQueue;
import java.lang.Math;

public class HelloWorld {    
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    TreeNode constructBST(int[] nums, int left, int right) {
        if (left>right) return null;
        int mid=left+(right-left)/2;
        TreeNode node=new TreeNode(nums[mid]);
        node.left=constructBST(nums, left, mid-1);
        node.right=constructBST(nums, mid+1, right);
        return node;
    }
    
    public TreeNode sortedArrayToBST(int[] nums) {
        return constructBST(nums, 0, nums.length-1);
    }
     
    public static void main(String[] args) {
    }
}

File: cpp/convert_sorted_array_to_binary_search_tree.cpp
108. Convert Sorted Array to Binary Search Tree  QuestionEditorial Solution  My Submissions
Total Accepted: 83034 Total Submissions: 216537 Difficulty: Medium
Given an array where elements are sorted in ascending order, convert it to a height balanced BST.

Hide Company Tags Airbnb
Hide Tags Tree Depth-first Search
Hide Similar Problems (M) Convert Sorted List to Binary Search Tree

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* convert(vector<int>& nums, int start, int end) {
        if (start>end) return NULL;
        int mid=start+(end-start)/2;
        TreeNode* n=new TreeNode(nums[mid]);
        n->left=convert(nums, start, mid-1);
        n->right=convert(nums, mid+1, end);
        return n;
    }
    
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return convert(nums, 0, nums.size()-1);
    }
};

